subsys/retention/retention.c

```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_retention
+
+#include <string.h>
+#include <sys/types.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/crc.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/retained_mem.h>
+#include <zephyr/retention/retention.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retention, CONFIG_RETENTION_LOG_LEVEL);
+
+#define DATA_VALID_VALUE 1
+
+enum {
+	CHECKSUM_TYPE_NONE = 0,
+	CHECKSUM_TYPE_CRC8,
+	CHECKSUM_TYPE_CRC16,
+	CHECKSUM_TYPE_UNUSED,
+	CHECKSUM_TYPE_CRC32,
+};
+
+struct retention_data {
+	bool header_written;
+#ifdef CONFIG_RETENTION_MUTEXES
+	struct k_mutex lock;
+#endif
+};
+
+struct retention_config {
+	const struct device *parent;
+	size_t offset;
+	size_t size;
+	size_t reserved_size;
+	uint8_t checksum;
```

Comment:
Should this parameter be renamed `checksum_len` or `checksum_size` it would make some code reading a bit easier like `ssize_t data_size = config->size - config->checksum;`

Reply:
Fixed